apps/system/js/launcher.js

```diff
@@ -0,0 +1,147 @@
+        this.readFtu(),
+        this.readSetting('homescreen.manifestURL'),
+        this.readSetting('ftu.manifestURL'),
+        this.readSetting('deviceinfo.os'),
+        this.readSetting('deviceinfo.previous_os'),
+        this.readSetting('lockscreen.enabled'),
+        this.readSetting('wallpaper.image'),
+        this.readSetting('wallpaper.image.valid')
+      ]).then(function(results) {
+        var [ftuEnabled, homescreenManifestURL, ftuManifestURL,
+             osVersion, previousOsVersion, lockscreenEnabled,
+             wallpaper, wallpaperValid] = results;
+        this.lockscreenEnabled = lockscreenEnabled;
+        this.wallpaper = wallpaper;
+        this.wallpaperValid = wallpaperValid;
+        this.debug('Ftu enabled = ' + ftuEnabled +
+          '; ftuManifestURL = ' + ftuManifestURL +
+          '; osVersion = ' + osVersion +
+          '; previousOsVersion = ' + previousOsVersion +
+          '; homescreenManifestURL = ' + homescreenManifestURL +
+          '; lockscreenEnabled = ' + lockscreenEnabled);
+        if (this.checkUpgrading(osVersion, previousOsVersion)) {
+          this.debug('upgrading boot');
+          this.launchFtuThenHomescreen(ftuManifestURL,
+            homescreenManifestURL);
+        } else {
+          if (ftuEnabled !== false && ftuManifestURL) {
+            this.debug('clean boot');
+            this.launchFtuThenHomescreen(ftuManifestURL,
+              homescreenManifestURL);
+          } else {
+            this.debug('normal boot');
+            if (lockscreenEnabled) {
+              this.launchLockscreenThenHomescreen(homescreenManifestURL);
+            } else {
+              this.launchHomescreenAndStandbyLockscreen(homescreenManifestURL);
+            }
+          }
+        }
+      }.bind(this));
+    },
+    justUpgraded: function() {
+      return this._isUpgrade;
+    },
+    checkUpgrading: function(prev, curr) {
+      var isUpgrade = false;
+      // dont treat lack of previous version info as an upgrade
+      if (prev && curr) {
+        isUpgrade = curr.major > prev.major || curr.minor > prev.minor;
+      }
+      this._isUpgrade = isUpgrade;
+      return isUpgrade;
+    },
+    launchHomescreenAndStandbyLockscreen: function(homescreenManifestURL) {
+      Promise.all([
+        // We still need to tell FtuLauncher to skip to process some tasks.
+        this.service.request('FtuLauncher:skip'),
+        this.service.request('WallpaperManager:initializeWallpaper',
+          this.wallpaper, this.wallpaperValid),
+        this.service.request('LockScreenLauncher:standby'),
+        this.service.request('HomescreenLauncher:launch',
+          homescreenManifestURL, true).then(() => {
+            this.service.request('LogoManager:animatePoweronLogo');
+            this.scheduler.release();
+          })
+      ]).catch((err) => {
+        console.error(err);
+      });
+    },
+    launchLockscreenThenHomescreen: function(homescreenManifestURL) {
+      Promise.all([
```

Comment:
Please explain why you would like to start a new chain in a comment above this line.